File: src/plnflr/engine/tile.py
"""Centered tile grid clipped to the inset polygon."""

from __future__ import annotations

from plnflr.domain.models import (
    LayoutPlan,
    LayoutRules,
    Piece,
    Room,
    TileSpec,
    Vertex,
)
from plnflr.engine.bom import make_bom
from plnflr.engine.clip import area_mm2, intersect_rect, ring_to_tuple
from plnflr.engine.expansion import bbox, inset_room, resolve_gap_mm
from plnflr.engine.instructions import instructions_for


def layout_tiles(room: Room, spec: TileSpec, rules: LayoutRules) -> LayoutPlan:
    gap = resolve_gap_mm(room, rules)
    inner = inset_room(room, gap)
    min_x, min_y, max_x, max_y = bbox(inner)
    inner_w = max_x - min_x
    inner_h = max_y - min_y
    pitch_x = spec.length_mm + spec.grout_mm
    pitch_y = spec.width_mm + spec.grout_mm
    cols = max(1, (inner_w + spec.grout_mm) // pitch_x)
    rows = max(1, (inner_h + spec.grout_mm) // pitch_y)
    used_w = cols * spec.length_mm + (cols - 1) * spec.grout_mm
    used_h = rows * spec.width_mm + (rows - 1) * spec.grout_mm
    # If remainder on an edge is too small, add a column/row so cuts grow.
    rem_x = inner_w - used_w
    rem_y = inner_h - used_h
    if 0 < rem_x < rules.min_tile_cut_mm:
        cols += 1
        used_w = cols * spec.length_mm + (cols - 1) * spec.grout_mm
        rem_x = inner_w - used_w
    if 0 < rem_y < rules.min_tile_cut_mm:
        rows += 1
        used_h = rows * spec.width_mm + (rows - 1) * spec.grout_mm
        rem_y = inner_h - used_h
    origin_x = min_x + rem_x // 2
    origin_y = min_y + rem_y // 2
    outer_path = ring_to_tuple(inner.outer)
    hole_paths = tuple(ring_to_tuple(h) for h in inner.holes)
    pieces: list[Piece] = []
    board = 0
    order = 0
    for row in range(rows):
        for col in range(cols):
            x0 = origin_x + col * pitch_x
            y0 = origin_y + row * pitch_y
            x1 = x0 + spec.length_mm
            y1 = y0 + spec.width_mm
            rect = ((x0, y0), (x1, y0), (x1, y1), (x0, y1))
            clipped = intersect_rect(rect, outer_path, hole_paths)
            if not clipped:
                continue
            board += 1
            for poly in clipped:
                order += 1
                kind = "tile_full" if len(poly) == 4 else "tile_cut"
                pieces.append(
                    Piece(
                        piece_id=f"t{row}-{col}-{order}",
                        row_index=row,
                        geometry=tuple(Vertex(x, y) for x, y in poly),
                        kind=kind,
                        source_board=board,
                        install_order=order,
                        length_mm=spec.length_mm,
                        width_mm=spec.width_mm,
                    )
                )
    net = area_mm2(outer_path) - sum(area_mm2(h) for h in hole_paths)
    bom = make_bom(
        pieces=len(pieces),
        full_boards=board,
        boards_per_pack=spec.tiles_per_pack,
        area_net_mm2=net,
        board_area_mm2=spec.length_mm * spec.width_mm,
    )
    return LayoutPlan(
        room=room,
        gap_mm=gap,
        inset=inner,
        direction="along_x",
        pieces=tuple(pieces),
        bom=bom,
        warnings=(),
        rationale_pl="Siatka płytek wycentrowana na bbox, przycięta do obrysu.",
        rows_instruction_pl=instructions_for(
            tuple(pieces), gap_mm=gap, has_holes=bool(room.holes)
        ),
    )

File: Kernel/Sources/PlnFlrLayout/Tile.swift
public func layoutTiles(
    _ room: Room,
    _ spec: TileSpec,
    _ rules: LayoutRules,
    windows: [Opening] = []
) throws -> LayoutPlan {
    let gap = try resolveGapMm(room, rules)
    let inner = try insetRoom(room, gapMm: gap)
    let angle = ((rules.angleDeg % 360) + 360) % 360
    let origin = originOf(inner)
    let gridRoom = angle == 0 ? inner : try rotateRoom(inner, angleDeg: angle, origin: origin)
    var pieces = try layTiles(gridRoom, spec, rules)
    if angle != 0 {
        pieces = pieces.map { piece in
            var copy = piece
            copy.geometry = rotatePoints(piece.geometry, angleDeg: -angle, origin: origin)
            return copy
        }
    }
    let net = areaMm2(inner.outer.vertices)
        - inner.holes.reduce(0) { $0 + areaMm2($1.vertices) }
    let bom = makeBom(
        pieces: pieces.count,
        fullBoards: pieces.map(\.sourceBoard).max() ?? 0,
        boardsPerPack: spec.tilesPerPack,
        areaNetMm2: net,
        boardAreaMm2: spec.lengthMm * spec.widthMm,
        label: "Płytki",
        kind: .tile
    )
    let angleNote = angle == 0 ? "" : " Kąt \(angle)°."
    return LayoutPlan(
        angleDeg: angle,
        bom: bom,
        boms: [bom],
        direction: .alongX,
        divider: nil,
        gapMm: gap,
        inset: inner,
        pieces: pieces,
        rationalePl: "Siatka płytek wycentrowana na bbox, przycięta do obrysu.\(angleNote)",
        room: room,
        rowsInstructionPl: instructionsFor(pieces, gapMm: gap, hasHoles: !room.holes.isEmpty),
        splitAtMm: nil,
        splitAxis: nil,
        warnings: [],
        windows: windows
    )
}

private func layTiles(_ inner: Room, _ spec: TileSpec, _ rules: LayoutRules) throws -> [Piece] {
    let bounds = bbox(inner)
    let innerW = bounds.maxX - bounds.minX
    let innerH = bounds.maxY - bounds.minY
    let pitchX = spec.lengthMm + spec.groutMm
    let pitchY = spec.widthMm + spec.groutMm
    var cols = max(1, (innerW + spec.groutMm) / pitchX)
    var rows = max(1, (innerH + spec.groutMm) / pitchY)
    var usedW = cols * spec.lengthMm + (cols - 1) * spec.groutMm
    var usedH = rows * spec.widthMm + (rows - 1) * spec.groutMm
    var remX = innerW - usedW
    var remY = innerH - usedH
    if remX > 0, remX < rules.minTileCutMm {
        cols += 1
        usedW = cols * spec.lengthMm + (cols - 1) * spec.groutMm
        remX = innerW - usedW
    }
    if remY > 0, remY < rules.minTileCutMm {
        rows += 1
        usedH = rows * spec.widthMm + (rows - 1) * spec.groutMm
        remY = innerH - usedH
    }
    let originX = bounds.minX + remX / 2
    let originY = bounds.minY + remY / 2
    var pieces: [Piece] = []
    var board = 0
    var order = 0
    for row in 0..<rows {
        for col in 0..<cols {
            let x0 = originX + col * pitchX
            let y0 = originY + row * pitchY
            let x1 = x0 + spec.lengthMm
            let y1 = y0 + spec.widthMm
            let rect = [Vertex(x0, y0), Vertex(x1, y0), Vertex(x1, y1), Vertex(x0, y1)]
            let clipped = try intersectRect(
                rect,
                outer: inner.outer.vertices,
                holes: inner.holes.map(\.vertices)
            )
            guard !clipped.isEmpty else { continue }
            board += 1
            for poly in clipped {
                order += 1
                pieces.append(
                    Piece(
                        geometry: poly,
                        installOrder: order,
                        kind: poly.count == 4 ? .tileFull : .tileCut,
                        lengthMm: spec.lengthMm,
                        pieceId: "t\(row)-\(col)-\(order)",
                        rowIndex: row,
                        sourceBoard: board,
                        widthMm: spec.widthMm
                    )
                )
            }
        }
    }
    return pieces
}
